packages/dd-trace/src/appsec/downstream_requests.js

```diff
@@ -0,0 +1,252 @@
+'use strict'
+
+const web = require('../plugins/util/web')
+const { updateRaspRuleMatchMetricTags } = require('./telemetry')
+const {
+  HTTP_OUTGOING_METHOD,
+  HTTP_OUTGOING_HEADERS,
+  HTTP_OUTGOING_RESPONSE_STATUS,
+  HTTP_OUTGOING_RESPONSE_HEADERS,
+  HTTP_OUTGOING_RESPONSE_BODY
+} = require('./addresses')
+
+const KNUTH_FACTOR = 11400714819323199488n // eslint-disable-line unicorn/numeric-separators-style
+const UINT64_MAX = (1n << 64n) - 1n
+
+let config
+let globalRequestCounter
+let perRequestBodyAnalysisCount
+let perRequestDownstreamAnalysisCount
+
+function enable (_config) {
+  config = _config
+  globalRequestCounter = 0n
+  perRequestBodyAnalysisCount = new WeakMap()
+  perRequestDownstreamAnalysisCount = new WeakMap()
+}
+
+function disable () {
+  config = null
+  globalRequestCounter = 0n
+  perRequestBodyAnalysisCount = new WeakMap()
+  perRequestDownstreamAnalysisCount = new WeakMap()
+}
+
+/**
+ * Determines whether the current downstream request/responses bodies should be sampled for analysis.
+ * @param {import('http').IncomingMessage} req inbound request associated with the downstream call.
+ * @returns {boolean} true when the downstream response body should be captured.
+ */
+function shouldSampleBody (req) {
+  globalRequestCounter = (globalRequestCounter + 1n) & UINT64_MAX
+
+  const {
+    maxDownstreamRequestBodyAnalysis,
+    downstreamRequestBodyAnalysisSampleRate
+  } = config.appsec.apiSecurity
+
+  const currentCount = perRequestBodyAnalysisCount.get(req) || 0
+  if (currentCount >= maxDownstreamRequestBodyAnalysis) {
+    return false
+  }
+
+  const samplingRate = Math.min(Math.max(downstreamRequestBodyAnalysisSampleRate, 0), 1)
+
+  const hashed = (globalRequestCounter * KNUTH_FACTOR) % UINT64_MAX
+  const threshold = BigInt(Math.round(samplingRate * Number(UINT64_MAX)))
```

Comment:
isn't a bit weird that we are converting BigInt to number to make an operation to convert it againt to BigInt? Aren't we loosing the whole point of using BigInt?

```suggestion
  // replace 1000n with the accuraccy that we want to maintain
  const threshold = BigInt(Math.round(samplingRate * 1000n)) * UINT64_MAX / 1000n
```